### ISSUE TITLE
Multithreading for TDVP1

### ISSUE THREAD

**Issue by HumpyBlumpy** [maartenvd/MPSKit.jl#47]
Hi Maarten,

I have a question regarding how to best make use of multithreading. I read the section in the documentation which helped, but I think it does not cover my specific case.
The bulk of my simulation is time-evolution of a MPSComoving using 1-site TDVP. As far as I understand, there is nothing parallelizable  at the "high-level" in MPSKit (or in general for 1-site TDVP) for this case. So the only improvement could come from low-level BLAS operations. Hence, I start Julia with only 1 julia thread, and instead set OPENBLAS_NUM_THREADS to my max number of cores (which is pretty high on our cluster = 128). This does not seem however to lead to an improvement, so I wonder what is the proper way of making use of multithreading in this case?
I suppose it boils down to how TensorKit, KrylovKit, TensorOperations handle it.
Most of them mention the Strided package. 
So is the better strategy to start Julia --threads 128 and add 
```
Strided.enable_threads()
Strided.enable_threaded_mul()
```
and set BLAS threads to 1?

Thanks,
Ron

**Comment by HumpyBlumpy** [comment]
I managed to obtain a 2x speedup with the strided method above, and 1 blas thread. Too many julia threads seem to actually slow things down. ~10 seems to to be the sweet spot for my application.

**Comment by maartenvd** [comment]
Hi HumpyBlumpy,

specifically for TDVP the parallelization at the level of MPSKit is quite limited - it can only really parallelizes over the dense sub-blocks in the mpo hamiltonian, and there are typically not that many. You may get slightly better performance from mpskit by replacing every occurance of WorkStealingEx() by ThreadedEx() in the source code (like here https://github.com/maartenvd/MPSKit.jl/blob/8638cf92d345bc5cd7781ccc3507442167b088f7/src/algorithms/derivatives.jl#L67 ), I've not benchmarked this yet, and ideally would like to make all of this configurable somewhere.

At the tensorkit level, I don't think there's much parallization happening either. The relevant line for symmetric tensormaps (without symmetries there's nothing you can gain anyway) is https://github.com/Jutho/TensorKit.jl/blob/14d2490ca425ec08ffb9f267179f25070debc5dd/src/tensors/linalg.jl#L250
the "for c in blocksectors" can in principle be made an "@threads for c in blocksectors". 

On the strided level, the story gets more complicated. OpenBlas threads (the default blas that julia uses) does not co-operate at all with julia threads, so whatever julia-threading-advantage we try to gain gets lost because of openblas. There are two workarounds - one is to switch to MKL, which suposedly behaves differently. The other is to set blas threading to 1, but then the multiplication of large dense matrices are also done using a single thread. Strided has this enable_threaded_mul option (and also a number-of-strided-threads setting) to nevertheless attempt to parallelize large dense matrices.

Long story short, if you stay with openblas (which I also what I'm doing) then best performance is typically achieved with blas_threads=1 and playing with both the number of julia threads and strided threads.

If you have a small benchmark script for your usecase (system size, hamiltonian, symmetries), then I can also look at it from my end and see if there are any low hanging fruits :)

**Comment by HumpyBlumpy** [comment]
Thanks for the detailed explanation!
Given that I dont have any symmetries, it seems that strided is the way to go, and I got a ~2x speed up which seems decent.
